libbeat/publisher/queue/diskqueue/serialize.go

```diff
@@ -84,16 +92,23 @@ func (e *eventEncoder) reset() {
 }
 
 func (e *eventEncoder) encode(evt interface{}) ([]byte, error) {
-	event, ok := evt.(publisher.Event)
-	if !ok {
-		// In order to support events of varying type, the disk queue needs
-		// to know how to encode them. When we decide to do this, we'll need
-		// to add an encoder to the settings passed in when creating a disk
-		// queue. See https://github.com/elastic/elastic-agent-shipper/issues/41.
-		// For now, just return an error.
-		return nil, fmt.Errorf("disk queue only supports publisher.Event")
+	switch v := evt.(type) {
+	case publisher.Event:
+		if e.serializationFormat != SerializationCBOR {
+			return nil, fmt.Errorf("incompatible serialization for type %T", v)
+		}
+		return e.encode_publisher_event(v)
+	case *messages.Event:
+		if e.serializationFormat != SerializationProtobuf {
+			return nil, fmt.Errorf("incompatible serialization for type %T", v)
+		}
```

Comment:
[Suggestion]

On both cases there is only one possible serialisation format, we could already add this information in the error message:
```suggestion
		if e.serializationFormat != SerializationCBOR {
			return nil, fmt.Errorf("incompatible serialization for type %T. Only CBOR is supported", v)
		}
		return e.encode_publisher_event(v)
	case *messages.Event:
		if e.serializationFormat != SerializationProtobuf {
			return nil, fmt.Errorf("incompatible serialization for type %T. Only Protobuf is supported", v)
		}
```